icmp.go

```diff
@@ -0,0 +1,495 @@
+	if te, ok := msg.Body.(*icmp.TimeExceeded); ok {
+		data = te.Data
+	} else if du, ok := msg.Body.(*icmp.DstUnreach); ok {
+		data = du.Data
+	} else {
+		return nil, fmt.Errorf("Unknown message type: %v", msg.Type)
+	}
+	// data should now have the IP header of the original message plus at least
+	// 8 bytes of the original message (which is, at least, the UDP header)
+	srcHeader, err := icmp.ParseIPv4Header(data)
+	if err != nil {
+		return nil, err
+	}
+	udpHeader := data[srcHeader.Len:]
+	if len(udpHeader) < 8 {
+		return nil, fmt.Errorf("source udp header too short: %d", len(udpHeader))
+	}
+	srcPort := binary.BigEndian.Uint16(udpHeader[0:2])
+	dstPort := binary.BigEndian.Uint16(udpHeader[2:4])
+	var name string
+	names, _ := net.LookupAddr(replyHeader.Src.String())
+	if len(names) > 0 {
+		name = names[0]
+	} else {
+		name = replyHeader.Src.String()
+	}
+	id := c.packetID(srcHeader.ID, srcHeader.Dst, int(srcPort), int(dstPort))
+	return &Hop{
+		Src: Addr{
+			IP: srcHeader.Src,
+		},
+		Dst: Addr{
+			IP: srcHeader.Dst,
+		},
+		Node: Addr{
+			Host: name,
+			IP:   replyHeader.Src,
+		},
+		Received: now,
+		ID:       id,
+		icmpType: icmpType,
+	}, nil
+}
+
+func (c *Continuous) readICMP(ctx context.Context, hops chan *Hop) error {
+	defer close(hops)
+	localAddr, err := findAddress()
+	if err != nil {
+		return err
+	}
+	// set up the listening socket for incoming ICMP packets
+	recvSocket, err := syscall.Socket(syscall.AF_INET, syscall.SOCK_RAW, syscall.IPPROTO_ICMP)
+	if err != nil {
+		// this is fatal, because we need to be able to read the responses
+		return err
+	}
+	defer syscall.Close(recvSocket)
+	// TODO: do we need a receive timeout if we're running continuously?
+	// we do, if we want ctx.Done() to exit the recvfrom loop, otherwise it blocks forever
+	// once we stop sending requests
+	// tv := syscall.NsecToTimeval(int64(c.Timeout))
+	// err = syscall.SetsockoptTimeval(recvSocket, syscall.SOL_SOCKET, syscall.SO_RCVTIMEO, &tv)
+	// if err != nil {
+	// 	// this one isn't really fatal, but we'll treat it as so, because it's a really bad sign
+	// 	return err
+	// }
+	err = syscall.Bind(recvSocket, &syscall.SockaddrInet4{Port: c.FirstPort, Addr: localAddr})
+	if err != nil {
+		// again, this is fatal
+		return err
+	}
+	for ctx.Err() == nil {
+		var p = make([]byte, 100)
+		n, _, err := syscall.Recvfrom(recvSocket, p, 0)
+		now := time.Now().UTC()
+		if err != nil {
+			c.logger.WithError(err).Error("Recvfrom error")
+			time.Sleep(10 * time.Millisecond)
+		} else {
+			// extractMessage has to do reverse DNS lookups, which (at least on my machine) seem to get throttled,
+			// and this messes up the elapsed time calculation (by backing up the Recvfrom loop). So, we do it
+			// in its own goroutine.
+			go func(now time.Time, p []byte, n int, err error) {
+				if err != nil {
+					return
+				}
+				msg, err := c.extractMessage(p[:n], now)
+				if err != nil {
+					c.logger.WithError(err).Error("extractMessage error")
```

Comment:
Is this error generally caused by the throttling, and that's why it's logged rather than returning to the caller via errgroup or similar?

Reply:
The way the code is constructed, that func is not supposed to return, except for fatal errors, so I have to log it locally (or introduce an error channel). 
The extractMessage func doesn't return an error for a failed IP address lookup, because that happens pretty frequently. Any errors would be for parsing the packets, which is worth logging, because it's probably a bug.